app/forms/declaration/DocumentsProduced.scala

```diff
@@ -27,52 +28,104 @@ case class DocumentsProduced(
+  val documentTypeCodeKey = "documentTypeCode"
+  val documentIdentifierKey = "documentIdentifier"
+  val documentPartKey = "documentPart"
+  val documentStatusKey = "documentStatus"
+  val documentStatusReasonKey = "documentStatusReason"
+  val issuingAuthorityNameKey = "issuingAuthorityName"
+  val dateOfValidityKey = "dateOfValidity"
+  val measurementUnitKey = "measurementUnit"
+  val documentQuantityKey = "documentQuantity"
 
-  def form(): Form[DocumentsProduced] = Form(mapping)
+  val mapping = Forms
+    .mapping(
+      documentTypeCodeKey -> optional(
+        text().verifying("supplementary.addDocument.documentTypeCode.error", hasSpecificLength(4) and isAlphanumeric)
+      ),
+      documentIdentifierKey -> optional(
+        text().verifying("supplementary.addDocument.documentIdentifier.error", isAlphanumeric and noLongerThan(30))
+      ),
+      documentPartKey -> optional(
+        text().verifying("supplementary.addDocument.documentPart.error", isAlphanumeric and noLongerThan(5))
+      ),
+      documentStatusKey -> optional(
+        text().verifying("supplementary.addDocument.documentStatus.error", noLongerThan(2) and isAllCapitalLetter)
+      ),
+      documentStatusReasonKey -> optional(
+        text().verifying("supplementary.addDocument.documentStatusReason.error", noLongerThan(35) and isAlphanumeric)
+      ),
+      issuingAuthorityNameKey -> optional(
+        text()
+          .verifying(
+            "supplementary.addDocument.issuingAuthorityName.error.length",
+            noLongerThan(issuingAuthorityNameMaxLength)
+          )
+      ),
+      dateOfValidityKey -> optional(Date.mapping),
+      measurementUnitKey -> optional(
+        text()
+          .verifying(
+            "supplementary.addDocument.measurementUnit.error.length",
+            hasSpecificLength(measurementUnitMaxLength)
+          )
+          .verifying("supplementary.addDocument.measurementUnit.error.specialCharacters", isAlphanumeric)
+      ),
+      documentQuantityKey ->
+        optional(
+          bigDecimal
+            .verifying(
+              "supplementary.addDocument.documentQuantity.error.precision",
+              _.precision <= documentQuantityMaxLength
+            )
+            .verifying(
+              "supplementary.addDocument.documentQuantity.error.scale",
+              _.scale <= documentQuantityMaxDecimalPlaces
+            )
+            .verifying("supplementary.addDocument.documentQuantity.error", _ >= 0)
+        )
+    )(DocumentsProduced.apply)(DocumentsProduced.unapply)
+    .verifying("supplementary.addDocument.error.documentIdentifierAndPart", validateDocumentIdentifierAndPart(_))
+    .verifying(
+      "supplementary.addDocument.error.measurementUnitAndQuantity",
+      validateMeasurementUnitAndDocumentQuantity(_)
+    )
 
-  def fromJson(str: JsValue): DocumentsProduced = Json.fromJson(str).get
+  private def validateDocumentIdentifierAndPart(doc: DocumentsProduced): Boolean =
```

Comment:
this dependent validation can be placed as a generic validation across forms with something similar 
def requireAllDependantFields[T](primary: T => Option[_])(fs: (T => Option[_])*): T => Boolean =
    t => primary(t).fold(true)(_ => fs.forall(f => f(t).nonEmpty))

Reply:
This is good idea. However, this is code refactor so I will work on that during the next PR.